### ISSUE TITLE
Copy and paste from wine to terminal

### ISSUE THREAD

**Issue by syco** [asbru-cm/asbru-cm#673]
Hi, I'm having this weird issue with copy and paste, hopefully I'm not the only one..

**Describe the bug**
When I copy from an application running in wine I can't paste in asbru terminal (I tried both "ctrl + shift + v" and "right click + paste").
But If I open geany/terminator/firefox/netbeans (pick one), I can paste, copy again, and only then I'm able to paste in asbru.

**Environment (please complete the following information):**
 - OS: openSUSE Thumbleweed
 - Ásbrú Version: v6.2.2

Is anyone else experiencing this issue?
Thanks

**Comment by gfrenoy** [comment]
Did not try but as a general rule, Ásbrú is using the PRIMAY clipboard (see [definition](https://specifications.freedesktop.org/clipboards-spec/clipboards-latest.txt)).

Can you double check that the copied text from Wine is correctly available in that primary clipboard ?

Not sure for openSUSE but on my Ubuntu, `xsel -p` would give me the content of the primary clipboard.

**Comment by syco** [comment]
Copy from wine doesn't go to the primary, `xsel -p` is empty.
It goes however to the clipboard `xsel -b`.

Reading the definitions I don't think using always the primary (even for explicit copy and paste from buttons and shourtcuts) is the right thing to do.
```txt
 - selecting but with no explicit copy should only set PRIMARY,
   never CLIPBOARD

 - middle mouse button should paste PRIMARY, never CLIPBOARD

 - explicit cut/copy commands (i.e. menu items, toolbar buttons)
   should always set CLIPBOARD to the currently-selected data (i.e.
   conceptually copy PRIMARY to CLIPBOARD)

 - explicit cut/copy commands should always set both CLIPBOARD and
   PRIMARY, even when copying doesn't involve a selection (e.g. a
   "copy url" -option which explicitly copies an url without the
   url being selected first)

 - explicit paste commands should paste CLIPBOARD, not PRIMARY

 - a selection becoming unselected should never unset PRIMARY
```

**Comment by gfrenoy** [comment]
> I don't think using always the primary is the right thing to do

See #652 ; we're having an issue with CLIPBOARD ...

Not sure what is the best strategy here since nothing is really 100% working :(

**Comment by syco** [comment]
I see, but Shamanky's problem is with shift+insert, so maybe we can do what terminator does and make shift+insert and ctrl+shift+v do different things.

This is how it works for terminator:
highlight text : copy to PRIMARY
shift+insert : paste the PRIMARY
ctrl+shift+c : copy to CLIPBOARD
ctrl+shift+v : paste from CLIPBOARD

**Comment by gfrenoy** [comment]
Interesting idea, indeed.  Thanks!

Shortcut keys are now customizable but we can certainly have two different actions to cope with the various scenarios.

First note: the changes I made (available in 6.2.2 and 6.3.0) are not released yet so it can also make sense to revert them and start from what we had in 6.2.1 and implement this suggestion instead.

If we do that, I'd guess we only have to implement "paste-primary" ; for the rest, it was already working like described.

If someone wants to start with this, here are few suggestions:

- Add new action in the `PACKeyBindings.pm` -> https://github.com/asbru-cm/asbru-cm/blob/loki/lib/PACKeyBindings.pm#L387
- Note: do not rename existing actions ; I'd suggest to have `paste` for CLIPBOARD and `paste-primary` for PRIMARY
- Implement the different actions in `PACTerminal.pm` --> https://github.com/asbru-cm/asbru-cm/blob/loki/lib/PACTerminal.pm#L1039
- Search for 'CLIPBOARD' in the remaining code to see when it has to be replace

**Comment by syco** [comment]
I was looking into this change, but there's many places where primary/clipboard would still be up for debate.
Like in the "Copy Connection Password" and the "Paste and delete" shortcuts.
I believe they should come off the clipboard, not the primary. But to avoid a flame war it might be better to make it configurable.

Also I'm a little confused by the `_pasteToVte` function and the `slow` parameter.

Here's what I've done so far: [Possible pull request](https://github.com/asbru-cm/asbru-cm/compare/loki...syco:loki?expand=1)
As you can see, I've also removed the `_pasteToVte` function and here's why:
```perl
        } elsif ($action eq 'paste') {
            my $txt = $$self{_GUI}{_VTE}->get_clipboard(Gtk3::Gdk::Atom::intern_static_string('PRIMARY'))->wait_for_text();
            $self->_pasteToVte($txt, $$self{_CFG}{'environments'}{$$self{_UUID}}{'send slow'});
```
Lines 1040/1041 in `PACTerminal.pm`, where the paste is handled, doesn't matter I use PRIMARY or CLIPBOARD, because the `send slow` is always 0. So `_pasteToVte` would always go to the "else" branch of the "if".
My solution is to remove `_pasteToVte` and use `paste_clipboard`, `paste_primary`, and `_vteFeedChild` directly.

**Comment by gfrenoy** [comment]
> I believe they should come off the clipboard, not the primary

In the current "official" version (6.2.1) ; this is the case.  I changed it for the next versions (6.2.2 or 6.3.0) but it is not too late to change our mind.  I'd really suggest to revert what I did and re-think about it (now that we have learned more things / are more clever ;-p).

> Also I'm a little confused by the _pasteToVte function and the slow parameter.

Option `send slow` corresponds to this UI option:

![image](https://user-images.githubusercontent.com/981546/89104685-7330b980-d41b-11ea-8e61-5060fcd32008.png)

Not sure if many are really using it and if it's safe to remove.  Especially because in some places of the code, it is forced to 1 so it must be useful for something, don't you think ?

**Comment by syco** [comment]
I reorganized my forked repo and had to delete my previous commits.
My proposed changes for the copy and paste are here now: [Copy and paste](https://github.com/asbru-cm/asbru-cm/compare/loki...syco:loki-feature-clipboard)

Regarding `_pasteToVte`:
since the original developer is long gone we can only make assumptions.
One is that _vteFeedChild wasn't available at the time, and he had to send the string one char at a time. In that case, putting pause between subsequent sends is a good thing, I had to do it sometimes in other languages.
Another possibility is that it is a workaround for some bug in _vteFeedChild, gtk2, or old vte.

But as it stands now I don't see any reason not to use _vteFeedChild with the whole string.
I would consider splitting a string and sending chars one by one bad practice since vte now has _vteFeedChild and that it seems to be working.
